crates/pathfinder/src/sequencer/builder.rs

```diff
@@ -0,0 +1,233 @@
+        self.with_gateway("gateway")
+    }
+
+    pub fn feeder_gateway(self) -> Request<'a, WithGateWay> {
+        self.with_gateway("feeder_gateway")
+    }
+
+    fn with_gateway(mut self, gateway: &str) -> Request<'a, WithGateWay> {
+        self.url
+            .path_segments_mut()
+            .expect("Base URL is valid")
+            .push(gateway);
+        Request {
+            url: self.url,
+            client: self.client,
+            marker: PhantomData::default(),
+        }
+    }
+}
+
+impl<'a> Request<'a, WithGateWay> {
+    pub fn add_transaction(self) -> Request<'a, WithMethod> {
+        self.with_method("add_transaction")
+    }
+
+    pub fn call_contract(self) -> Request<'a, WithMethod> {
+        self.with_method("call_contract")
+    }
+
+    pub fn get_block(self) -> Request<'a, WithMethod> {
+        self.with_method("get_block")
+    }
+
+    pub fn get_full_contract(self) -> Request<'a, WithMethod> {
+        self.with_method("get_full_contract")
+    }
+
+    pub fn get_class_by_hash(self) -> Request<'a, WithMethod> {
+        self.with_method("get_class_by_hash")
+    }
+
+    pub fn get_class_hash_at(self) -> Request<'a, WithMethod> {
+        self.with_method("get_class_hash_at")
+    }
+
+    pub fn get_storage_at(self) -> Request<'a, WithMethod> {
+        self.with_method("get_storage_at")
+    }
+
+    pub fn get_transaction(self) -> Request<'a, WithMethod> {
+        self.with_method("get_transaction")
+    }
+
+    pub fn get_transaction_status(self) -> Request<'a, WithMethod> {
+        self.with_method("get_transaction_status")
+    }
+
+    pub fn get_state_update(self) -> Request<'a, WithMethod> {
+        self.with_method("get_state_update")
+    }
+
+    pub fn get_contract_addresses(self) -> Request<'a, WithMethod> {
+        self.with_method("get_contract_addresses")
+    }
+
+    #[cfg(test)]
+    pub fn custom(self, method: &'static str) -> Request<'a, WithMethod> {
+        self.with_method(method)
+    }
+
+    fn with_method(mut self, method: &str) -> Request<'a, WithMethod> {
+        self.url
+            .path_segments_mut()
+            .expect("Base URL is valid")
+            .push(method);
+
+        Request {
+            url: self.url,
+            client: self.client,
+            marker: PhantomData::default(),
+        }
+    }
+}
+
+impl<'a> Request<'a, WithMethod> {
+    pub fn at_block<B: Into<BlockId>>(self, block: B) -> Self {
+        use std::borrow::Cow;
+
+        let block: BlockId = block.into();
+        let (name, value) = match block {
+            BlockId::Number(number) => ("blockNumber", Cow::from(number.0.to_string())),
+            BlockId::Hash(hash) => ("blockHash", hash.0.to_hex_str()),
+            // The name for these two could be either "blockNumber" or "blockHash".
+            BlockId::Latest => ("blockNumber", Cow::from("latest")),
+            BlockId::Pending => ("blockNumber", Cow::from("pending")),
```

Comment:
Hmm, interestingly `blockHash` still does not work with [`latest`](https://alpha4.starknet.io/feeder_gateway/get_block?blockHash=latest) or [`pending`](https://alpha4.starknet.io/feeder_gateway/get_block?blockHash=pending).

Reply:
wait so it has to be `blockNumber` and I just happened to choose the correct variant?

Reply:
Exactly :tada: :dart: .